java/source/infodynamics/measures/mixed/kraskov/MutualInfoCalculatorMultiVariateWithDiscreteKraskov.java

```diff
@@ -247,124 +461,130 @@ protected void computeNorms() {
+		// Perform new initialisations
+		miSurrogateCalculator.initialise(dimensions, base);
+		// Set new observations
+		miSurrogateCalculator.setObservations(continuousData, shuffledDiscreteData);
+		// Compute the MI
+		return miSurrogateCalculator.computeAverageLocalOfObservations();
 	}
 
-	public double computeAverageLocalOfObservations() throws Exception {
-		if (!tryKeepAllPairsNorms || (continuousData.length > MAX_DATA_SIZE_FOR_KEEP_ALL_PAIRS_NORM)) {
-			return computeAverageLocalOfObservationsWhileComputingDistances();
-		}
-		
-		// Postcondition: we'll compute the norms before the main loop,
-		//  unless they have already been computed:
-		
-		if (xNorms == null) {
-			computeNorms();
-		}
-		int N = continuousData.length; // number of observations
+  /**
+   * Calculate the average MI in nats.
+   */
+  public double computeAverageLocalOfObservations() throws Exception {
+    return computeFromObservations(false)[0];
+  }
+
+  /**
+   * <p>Computes the local values of the MI,
+   *  for each valid observation in the previously supplied observations
+   *  (with PDFs computed using all of the previously supplied observation sets).</p>
+   *  
+   * <p>If the samples were supplied via a single call such as
+   * {@link #setObservations(double[][], int[])},
+   * then the return value is a single time-series of local
+   * channel measure values corresponding to these samples.</p>
+   * 
+   * <p>Otherwise where disjoint time-series observations were supplied using several 
+   *  calls such as {@link #addObservations(double[][], int[])}
+   *  then the local values for each disjoint observation set will be appended here
+   *  to create a single "time-series" return array.</p>
+   * 
+   * @return the "time-series" of local MIs in nats
+   * @throws Exception
+   */
+  public double[] computeLocalOfPreviousObservations() throws Exception {
+    return computeFromObservations(true);
+  }
+
+  /**
+   * This protected method handles the computation of either the average or
+   * local MI (over parts of the total observations).
+   * 
+   * <p>The method returns:<ol>
+   *  <li>for (returnLocals == false), an array of size 1,
+   *      containing the average MI </li>
+   *  <li>for local MIs (returnLocals == true), the array of local MI values</li>
+   *  </ol>
+   * 
+   * @param returnLocals whether to return an array or local values, or else
+   *  sums of these values
+   * @return either the average MI, or array of local MI value, in nats not bits
+   * @throws Exception
+   */
+	protected double[] computeFromObservations(boolean returnLocals) throws Exception {
+
+    int[] cumcount = new int[base];
+    Arrays.fill(cumcount, 0);
+
 
 		// Count the average number of points within eps_x and eps_y
 		double averageDiGammas = 0;
 		double avNx = 0;
 		double avNy = 0;
 		
 		double testSum = 0.0; // Used for debugging prints
-		
+    int N = totalObservations;
+    double[] locals = new double[N];

```

Comment:
Perhaps only create locals[] if the boolean method argument is true (same as in the continuous KSG estimators). That would save allocating memory multiple times which may make a difference in large data sets and with many surrogates.

Reply:
Sure thing. Fixed in https://github.com/jlizier/jidt/pull/58/commits/25df49fa5a1806e1c132277ad730a04c9994ee16